### PR TITLE
Support for multiple tags

### PR DESCRIPTION
**Pull request by chergey**
I added support for multiple tags.

### CODE REVIEW

**Comment by webmat** [comment]
Thanks for taking the time to work on this.

Currently it's already possible to specify multiple tags, as mentioned in [the doc](https://github.com/logstash-plugins/logstash-output-loggly/blob/master/docs/index.asciidoc#plugins-outputs-loggly-tag).

What's the issue you're trying to fix? Interpolation like the following should work as well, even with multiple tags:

```
output {
  loggly { ... "tag" => "tag1,%{log_type}" }
}
```

Or have you encountered issues where this didn't work as expected?

**Comment by chcergey** [comment]
> output {
>   loggly { ... "tag" => "tag1,%{log_type}" }
> }
That didn't work for me.

**Comment by webmat** [comment]
The problem likely comes from your field `log_type ` not being present.

If I start Logstash with the following:

```bash
bin/logstash -e "input { stdin {} generator { count => 1 } }
  output {  loggly { key => \"$LOGGLY_KEY\" tag => 'logstash,tag_%{sequence}' } }
```

The generator plugin will generate one event with field `"sequence":0`. The event correctly appears in Loggly with tags `logstash` and `tag_0`.

However if I type another message, the `stdin` input doesn't set a `sequence` field. Since the failed interpolation leaves the `tag_%{sequence}` in place, and [this bit of code](https://github.com/logstash-plugins/logstash-output-loggly/blob/06c8d37/lib/logstash/outputs/loggly.rb#L146) comes down like a hammer, and replaces the whole `tag` attribute with the default tag. This is indeed not ideal.

**Comment by webmat** [comment]
However I would like it if you could do a few things in your PR, prior to merging this.

- [ ] Make sure to de-duplicate the tag list by calling `.uniq` on it, in case the list of tags ends up being "logstash,logstash,mylog", in case of interpolation problems.
- [ ] Please add tests for the various scenarios covered by this improved support for multiple tags. Here are some ideas:
  - [ ] Unit tests around scenarios of calling `prepare_meta` for different scenarios: no interpolation, interpolation of one tag, interpolation of one tag in a list of tags, a failed interpolation, duplicate tags are deduped.
    - [Here is an example](https://github.com/logstash-plugins/logstash-output-loggly/blob/master/spec/outputs/loggly_spec.rb#L129-L151) of unit tests directly calling `prepare_meta`, it may give you some ideas on how to get started.
  - [ ] Tests around the grouping of API calls based on different mixes of tags. Since the API requires the tag list to be part of the URI, and since the plugin accepts batches of events that can all have different lists of tags, we have to group the calls to do one bulk API call per pair of key + tag list.
    - [Here are the current tests](https://github.com/logstash-plugins/logstash-output-loggly/blob/master/spec/outputs/loggly_spec.rb#L156-L185) of splitting the calls in the proper key+tag list buckets. You can either add to those or create a new section.

If you're not familiar with RSpec, our testing framework, you can always check [this site](https://relishapp.com/rspec), the Expectations and Mocks sections will likely help you figure out how to start from our existing tests and add new ones.

**Comment by webmat** [comment]
@chergey Also, could you make sure you sign our Contributor License Agreement, please? https://www.elastic.co/contributor-agreement

**Comment by chergey** [comment]
@webmat, how about we remove default tag ('logstash') ? If no tag supplied or failed interpolation, then no tag will result, e.g. %{tag1},tag2 would be 'tag2'.

**Comment by webmat** [comment]
Yes that makes sense.

Make sure to still default to 'logstash' if ever the final tag is an empty string (one or more interpolations failed + no literal tag). An example to make sure I'm clear:

```
#  fail    fail
'%{tag1},%{tag2}'
```

Should end up with tag `logstash` :-)

**Comment by webmat** [comment]
In other words, inside the loop, just don't add the tag if it didn't interpolate correctly (e.g. if it still matches the regex).

And only after the loop do you check whether it's necessary to default to 'logstash' (if array is empty).

**Comment by chergey** [comment]
@webmat, why still default to 'logstash' ? We can send events without any tag (http://logs-01.loggly.com/inputs/TOKEN).

**Comment by webmat** [comment]
Ah I was not aware that the tag was optional. Their documentation doesn't clearly state whether the tag is required or can be omitted.

If that is so, I don't mind sending with no tag. Make sure that the URL will be adjusted accordingly (I would assume that "...TOKEN/tag/" with no tags won't work.

And document the new behaviour in the comments for the attribute and in docs/index.asciidoc. GitHub renders asciidoc just like markdown, so you can check that what you did is fine (even if GitHub doesn't render the includes & so on at the top of the file).

**Comment by chergey** [comment]
@webmat, please comment on my input.

File: lib/logstash/outputs/loggly.rb [hunk 1]
```diff
@@ -135,25 +136,27 @@ def receive(event)
...
-	
+    tags=@tag.split(",")
+    tag_array = []
+
+    tags.each do |t|
+      t = event.sprintf(t)
+      # For those cases where %{somefield} doesn't exist we don't include it
+      if !/%{\w+}/.match(t)
+        tag_array.push(t)
+      end
```
[the remark this answers is not in the document]
Comment (webmat) [reply lib/logstash/outputs/loggly.rb]: Also minor code convention nitpick. Avoid `if !` and use `unless` instead.

File: lib/logstash/outputs/loggly.rb [hunk 2]
```diff
@@ -173,7 +176,12 @@ def format_message(event)
   def send_batch(meta_events)
     split_batches(meta_events.compact).each_pair do |k, batch|
       key, tag = *k
-      url = "#{@proto}://#{@host}/bulk/#{key}/tag/#{tag}"
+      if tag != nil
+        url = "#{@proto}://#{@host}/bulk/#{key}/tag/#{tag}"
+      else
+        url = "#{@proto}://#{@host}/bulk/#{key}"
+      end
```
Comment (webmat) [inline lib/logstash/outputs/loggly.rb]: Once again, avoid `if !`. `unless...else` is not better, though. So perhaps use something like this instead?

```ruby
if tag.nil?
  url = "#{@proto}://#{@host}/bulk/#{key}"
else
  url = "#{@proto}://#{@host}/bulk/#{key}/tag/#{tag}"
end
```

File: spec/outputs/loggly_spec.rb [hunk 3]
```diff
@@ -60,9 +56,9 @@ def logger_for(plugin)
       output.receive(event)
     end
 
-    it 'should default tag to logstash if interpolated field for tag does not exist' do
+    it 'should send no tag to logstash if interpolated field for tag does not exist' do
```
Comment (webmat) [inline spec/outputs/loggly_spec.rb]: Initial wording (before your contribution) was a bit wonky. Could you remove the inaccurate "to logstash", and perhaps have this instead?

`it 'should have no tag if the interpolated field for the tag does not exist' do`

File: spec/outputs/loggly_spec.rb [hunk 4]
```diff
@@ -110,6 +106,46 @@ def logger_for(plugin)
       end
     end
 
+    context 'prepare tags' do
+      it 'failed interpolation' do
```
Comment (webmat) [inline spec/outputs/loggly_spec.rb]: Your `context` and `it` strings should read more like sentences that describe the situation you're setting up (context) and intended result (it).

Here's an example for the two lines above:

```ruby
context 'when computing the tags'
  it 'should not leave a failed interpolation in the tag list' do
```

then of course all following `it` blocks should be clarified to describe expected behaviour of each test case as well.

File: spec/outputs/loggly_spec.rb [hunk 5]
```diff
@@ -110,6 +106,46 @@ def logger_for(plugin)
...
+        expect(meta_event[:tag]).to eq('some_value1,some_value2')
+      end
+
+      it 'duplicate tags are deduped' do
+        config['tag']  =  'some_value,some_value'
+
+        meta_event = output.send :prepare_meta, event
+        expect(meta_event[:tag]).to eq('some_value')
+      end
+    end
```
Comment (webmat) [inline spec/outputs/loggly_spec.rb]: Good coverage of the various scenarios, however. Love it 💯

**Comment by chergey** [comment]
> You did specifically confirm that it's fine to send API calls without the tag/#{tag}, correct?
Yes, we have our own api in java that uses works just fine. Besides, I re-checked it myself with rest client.
> Declaring the variable outside the if is not necessary outside of control blocks like if. This is only necessary for blocks like do...end
You sure, it's ok ? Intelliji IDEA plugins complains.

**Comment by webmat** [comment]
Excellent. Once again, thanks a lot for your contribution!

I will merge and release the next version of the plugin.

**Comment by webmat** [comment]
Ah I have a small bit of housekeeping for you first, though.

I see you've signed our CLA with your email `@mail.ru`, but not with your email for the first commits `@haulmon.com`. Could you sign the CLA with the latter email as well, please? http://www.elasticsearch.org/contributor-agreement/

**Comment by chergey** [comment]
> I see you've signed our CLA with your email @mail.ru, but not with your email for the first commits @haulmon.com. Could you sign the CLA with the latter email as well, please? http://www.elasticsearch.org/contributor-agreement/
Done.

**Comment by webmat** [comment]
Thanks, I will be releasing this shortly!

**Comment by webmat** [comment]
@chergey Version 6.0.0 with your changes is finally out (after a few internal back & forths :-) ). Once again, thanks a lot for taking the time to contribute!

**Pull request merged**